vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,527 @@
-
-        tags = ["vcenter_server:{}".format(ensure_unicode(instance.get('name')))]
-        regexes = {
-            'host_include': instance.get('host_include_only_regex'),
-            'vm_include': instance.get('vm_include_only_regex'),
-        }
-        include_only_marked = is_affirmative(instance.get('include_only_marked', False))
-
-        # Discover hosts and virtual machines
-        server_instance = self._get_server_instance(instance)
-        use_guest_hostname = is_affirmative(instance.get("use_guest_hostname", False))
-        all_objs = self._get_all_objs(
-            server_instance, regexes, include_only_marked, tags, use_guest_hostname=use_guest_hostname
+    def refresh_metrics_metadata_cache(self):
+        """Request the list of counters (metrics) from vSphere and store them in a cache."""
+        self.log.info(
+            "Refreshing the metrics metadata cache. Collecting all counters metadata for collection_level=%d",
+            self.collection_level,
         )
-
-        self.mor_objects_queue.fill(i_key, dict(all_objs))
-        self.cache_config.set_last(CacheConfig.Morlist, i_key, time.time())
-
-    @trace_method
-    def _process_mor_objects_queue_async(self, instance, mors):
-        """
-        Process a batch of items popped from the objects queue by querying the available
-        metrics for these MORs and then putting them in the Mor cache
-        """
-        t = time.time()
-        i_key = self._instance_key(instance)
-        server_instance = self._get_server_instance(instance)
-        perfManager = server_instance.content.perfManager
-
-        # For non realtime metrics, we need to specifically ask which counters are available for which entity,
-        # so we call perfManager.QueryAvailablePerfMetric for each cluster, datacenter, datastore
-        # This should be okay since the number of such entities shouldn't be excessively large
-        for mor in mors:
-            mor_name = str(mor['mor'])
-            available_metrics = {m.counterId for m in perfManager.QueryAvailablePerfMetric(entity=mor["mor"])}
-            try:
-                self.mor_cache.set_metrics(i_key, mor_name, self._compute_needed_metrics(instance, available_metrics))
-            except MorNotFoundError:
-                self.log.error("Object '%s' is missing from the cache, skipping. ", ensure_unicode(mor_name))
+        t0 = Timer()
+        counters = self.api.get_perf_counter_by_level(self.collection_level)
+        self.gauge("datadog.vsphere.refresh_metrics_metadata_cache.time", t0.total(), tags=self.base_tags, raw=True)
+        self.log.info("Collected %d counters metadata in %.3f seconds.", len(counters), t0.total())
+
+        for mor_type in self.collected_resource_types:
+            allowed_counters = []
+            for c in counters:
+                metric_name = format_metric_name(c)
+                if metric_name in ALLOWED_METRICS_FOR_MOR[mor_type] and not is_metric_excluded_by_filters(
```

Comment:
What's the reason for restricting to metrics listed in `ALLOWED_METRICS_FOR_MOR`?

That's would mean, if the user want a metric not listed or there is a new metric from vsphere, a code change is needed.

Reply:
Yes that's a drawback but there's a reason.
vSphere gives us two ways to list metrics:
- either we run the `QueryPerfCounterByLevel` api call and we get a list of metrics ids and names (but not the resource they apply to)
- or we run a similar query on each individual resource.

The issue with the first approach is that we don't know the resources a given metric is valid for. That means that have around 500 metrics in total but if only 20 of them applies to datastores, we will run a ton of useless query (which is expensive).

WIth the second approach, we have to run one query per resource, this is not doable for big environment.

Reply:
There is an intermediate approach, where we first get a reference to one of each resource (one VM, one Host, one datastore etc.). And then for this random reference, we ask vCenter for the list of metrics available for this resource.

That would give us a good approximation in realtime of what metrics are available for each resource. But this would only give us a subset of available metrics for each resource.

Reply:
> or we run a similar query on each individual resource.

> WIth the second approach, we have to run one query per resource, this is not doable for big environment.

You mean it's too expensive to run (5 types of resources VirtualMachine, HostSystem, Datacenter, Datastore, ClusterComputeResource)?